### PR TITLE
Added support for YARN node label expressions

### PR DESCRIPTION
**Pull request by superbobry**
The node label expression can be defined either for the whole application

```python
skein.ApplicationSpec(node_label="gpu")
```

or for each service separately

```python
skein.Service(..., node_label="gpu")
```

### CODE REVIEW

File: java/src/main/java/com/anaconda/skein/ApplicationMaster.java [hunk 1]
```diff
@@ -576,8 +576,13 @@ public synchronized void initialize() throws IOException {
...
       Priority priority = newPriority(this);
-      ContainerRequest req = new ContainerRequest(service.getResources(),
-                                                  null, null, priority);
+      ContainerRequest req = new ContainerRequest(
+          service.getResources(),
+          null,
+          null,
+          priority,
+          true,
+          service.getNodeLabel().isEmpty() ? null : service.getNodeLabel());
```
Comment (jcrist) [inline java/src/main/java/com/anaconda/skein/ApplicationMaster.java]: IIUC this should also potentially inherit from the application wide node label. I'd probably propogate the application wide default down the services in `loadApplicationSpec` or in the `ServiceTracker` constructor.

**Comment by superbobry** [comment]
IIRC node labelling was introduced in Hadoop 2.7.X which is not available in CDH5. We have backported this feature to 2.6.X internally, but for the docker-compose, I guess we'd have to drop CDH to have the labelling tested. Are you OK with that?

File: java/src/main/java/com/anaconda/skein/ApplicationMaster.java (hunk 1 above)
Comment (superbobry) [reply java/src/main/java/com/anaconda/skein/ApplicationMaster.java]: Makes sense, will do.

**Comment by jcrist** [comment]
Hmmm. The methods appear to be available in 2.6.0 (see e.g. http://hadoop.apache.org/docs/r2.6.0/api/org/apache/hadoop/yarn/api/records/ApplicationSubmissionContext.html#setNodeLabelExpression(java.lang.String), and are found with `javap` on the test cdh5 install.

**Comment by jcrist** [comment]
[The IBM docs](https://developer.ibm.com/hadoop/2017/03/10/yarn-node-labels/) say it was introduced in 2.6.0, but not great to use until 2.8. I suspect that's mostly from the scheduler perspective, and the user api would be stable.

**Comment by superbobry** [comment]
Hmm, OK, thanks for the references. I will try to make it work in the hadoop-test-cluster. Stay tuned for updates.

**Comment by superbobry** [comment]
@jcrist do you think we can merge this PR given that the tests pass on a two-worker cluster? I can rebase on master.

**Comment by jcrist** [comment]
Yeah, apologies for the delay here. If you handle fixing the merge conflicts, I'll handle whatever else I see is needed to get this in. Thanks for being patient here.

**Comment by superbobry** [comment]
@jcrist I have merged with master (rebase was a bit more painful). I hope it's OK.

**Comment by jcrist** [comment]
Thanks for your patience @superbobry, I've pushed up some small changes:

- Removed the runtime tests. I wasn't happy with the hacks to get this to work with hadoop 2.6 (larger investigation showed these were just bugs in how hadoop 2.6 handled node labels), and the tests were *mostly* smoketests anyway. Since this is just forwarding parameters from the spec to the the container requests, I'm fine with testing by inspection only (though it's not ideal).
- Removed propagating the application level `node_label` ourselves. When I mentioned you should do that above I wasn't aware that YARN already does this internally. No need to duplicate work.
- Added docs

Merging, thanks again for all your contributions.

**Pull request merged**